src/librustc_mir/transform/cs_propagate.rs

```diff
@@ -0,0 +1,425 @@
+
+impl<'tcx> CsLattice<'tcx> {
+    fn insert(&mut self, key: &Lvalue<'tcx>, val: Either<'tcx>) {
+        // FIXME: HashMap has no way to insert stuff without cloning the key even if it exists
+        // already.
+        match *key {
+            // Do not bother with statics – global state.
+            Lvalue::Static(_) => {}
+            // I feel like this could be handled, but needs special care. For example in code like
+            // this:
+            //
+            // ```
+            // var.field = false;
+            // something(&mut var);
+            // assert!(var.field);
+            // ```
+            //
+            // taking a reference to var should invalidate knowledge about all the
+            // projections of var and not just var itself. Currently we handle this by not
+            // keeping any knowledge about projections at all, but I think eventually we
+            // want to do so.
+            Lvalue::Projection(_) => {},
+            _ => match self.values.entry(key.clone()) {
+                Entry::Vacant(e) => {
+                    e.insert(val);
+                }
+                Entry::Occupied(mut e) => {
+                    e.get_mut().join(val);
+                }
+            }
+        }
+    }
+    fn remove(&mut self, key: &Lvalue<'tcx>) -> Option<Either<'tcx>> {
+        self.values.remove(key)
+    }
+    fn top(&mut self, key: &Lvalue<'tcx>) {
+        self.insert(key, Either::Top);
+    }
+}
+
+impl<'tcx> Lattice for CsLattice<'tcx> {
+    fn bottom() -> Self { CsLattice { values: FnvHashMap() } }
+    fn join(&mut self, mut other: Self) -> bool {
+        // Calculate inteersection this way:
+        let mut changed = false;
+        // First, drain the self.values into a list of equal values common to both maps.
+        let mut common_keys = vec![];
+        for (key, mut value) in self.values.drain() {
+            match other.values.remove(&key) {
+                    // self had the key, but not other, so removing
+                None => {
+                    changed = true;
+                }
+                Some(ov) => {
+                    changed |= value.join(ov);
+                    common_keys.push((key, value));
+                }
+            }
+        }
+        // Now, put each common key with equal value back into the map.
+        for (key, value) in common_keys {
+            self.values.insert(key, value);
+        }
+        changed
+    }
+}
+
+struct CsTransfer<'a, 'tcx: 'a> {
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+}
+
+impl<'a, 'tcx> Transfer<'tcx> for CsTransfer<'a, 'tcx> {
+    type Lattice = CsLattice<'tcx>;
+    type TerminatorReturn = Vec<Self::Lattice>;
+
+    fn stmt(&self, s: &Statement<'tcx>, mut lat: Self::Lattice)
+    -> Self::Lattice
+    {
+        let StatementKind::Assign(ref lval, ref rval) = s.kind;
+        match *rval {
+            Rvalue::Use(Operand::Consume(ref nlval)) => {
+                lat.insert(lval, Either::Lvalue(nlval.clone()));
```

Comment:
It *was* primarily for move propagation, but otherwise it is a good idea to keep it `Either::Lvalue` here, because in case of code like `tmp1 = var1; tmp1 = 0;` it would correctly pick up the fact that the tmp1 is a constant. This wouldn’t work with `Top` because of how `join` operation with ⊤ works. Just removing the lval from the map here doesn’t work either, because then in a block with two incoming edges there could be a case where ⊤ is not produced, when it really should be. Finally, just making a `Either::Lvalue` (without the contained Lvalue) is also suboptimal, because you’d then have to produce `Either::Top` every time two such values are joined.